akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventsBySlicesFirehose.scala

```diff
@@ -0,0 +1,246 @@
+}
+
+class EventsBySlicesFirehose(system: ActorSystem[_]) extends Extension {
+  import EventsBySlicesFirehose._
+  private val firehoses = new ConcurrentHashMap[FirehoseKey, Firehose]()
+
+  def getFirehose(entityType: String, sliceRange: Range): Firehose =
+    firehoses.computeIfAbsent(FirehoseKey(entityType, sliceRange), key => createFirehose(key))
+
+  private def createFirehose(key: FirehoseKey): Firehose = {
+    implicit val sys: ActorSystem[_] = system
+
+    val bufferSize = 32 // FIXME config
+
+    val firehoseHub =
+      eventsBySlices[Any](
+        key.entityType,
+        key.sliceRange.min,
+        key.sliceRange.max,
+        TimestampOffset(Instant.now(), Map.empty),
+        firehose = true)
+        .map { env =>
+          // increment counter before the hub
+          firehoses.get(key) match {
+            case null     => throw new IllegalStateException(s"Firehose [$key] not created. This is a bug.")
+            case firehose => firehose.inCount.incrementAndGet()
+          }
+          env
+        }
+        .runWith(BroadcastHub.sink[EventEnvelope[Any]](bufferSize))
+
+    // FIXME hub will be closed when last consumer is closed, need to detect that and start again (difficult concurrency)?
+
+    new Firehose(firehoseHub, bufferSize)
+  }
+
+  // FIXME can this be in a ReadJournal
+  def eventsBySlices[Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset): Source[EventEnvelope[Event], NotUsed] = {
+
+    val firehose = getFirehose(entityType, minSlice to maxSlice)
+    val consumerId = UUID.randomUUID().toString
+    val catchupKillSwitch = KillSwitches.shared("catchupKillSwitch")
+    val catchup =
+      eventsBySlices[Event](entityType, minSlice, maxSlice, offset, firehose = false)
+        .map {
+          case env: EventEnvelope[Any] @unchecked => CatchupEventEnvelope(env)
+        }
+        .via(catchupKillSwitch.flow)
+
+    val consumerKillSwitch = KillSwitches.shared("consumerKillSwitch")
+
+    firehose.firehoseHub
+      .map { env =>
+        firehose.updateConsumerTracking(consumerId, timestampOffset(env).timestamp, consumerKillSwitch)
+        FirehoseEventEnvelope(env)
+      }
+      .merge(catchup)
+      .via(MergeSelector.flow(consumerId, catchupKillSwitch))
+      .map(_.asInstanceOf[EventEnvelope[Event]])
+      .via(consumerKillSwitch.flow)
+  }
+
+  // can be overridden in tests
+  @nowarn("msg=never used") // firehose param used in tests
+  protected def eventsBySlices[Event](
+      entityType: String,
+      minSlice: Int,
+      maxSlice: Int,
+      offset: Offset,
+      firehose: Boolean): Source[EventEnvelope[Event], NotUsed] = {
+    val queryPluginId = "akka.persistence.r2dbc.query"
+    PersistenceQuery(system)
+      .readJournalFor[EventsBySliceQuery](queryPluginId)
+      .eventsBySlices(entityType, minSlice, maxSlice, offset)
+  }
+
+}
+
+// FIXME maybe write as GraphStage instead
+object MergeSelector {
```

Comment:
Not sure if important but sure could save a bunch of allocs if it was one stage doing the merge knowing each upstream and keeping a mode state instead of having to tag/wrap each in a marker envelope